### PR TITLE
Ph/allow sync from whitelisted peers only

### PR DESCRIPTION
**Pull request by MithrilMan**
Actually it prevents from to sync with non whitelisted peer but doesn't force whitelisted to send legacy header, so with the commits as it is, you can't whitelist a C# node

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs [hunk 1]
```diff
@@ -35,6 +36,14 @@ public class ProvenHeadersConsensusManagerBehavior : ConsensusManagerBehavior
         private readonly IProvenBlockHeaderStore provenBlockHeaderStore;
         private readonly int lastCheckpointHeight;
         private readonly CheckpointInfo lastCheckpointInfo;
+        private readonly ConnectionManagerSettings connectionManagerSettings;
+
+        /// <summary>
+        /// Specify if the node is a gateway or not.
+        /// Gateway are used internally by Stratis to prevent network split during the transaction from Headers to ProvenHeaders protocol.
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: transition

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs [hunk 2]
```diff
@@ -184,6 +200,12 @@ protected override GetHeadersPayload BuildGetHeadersPayload()
         {
             INetworkPeer peer = this.AttachedPeer;
 
+            if (this.isGateway && !peer.IsWhitelisted())
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: Not sure if it's needed here. I mean the whole added block of code.

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs [hunk 3]
```diff
@@ -222,6 +244,12 @@ protected override GetHeadersPayload BuildGetHeadersPayload()
         /// <param name="headers">The headers.</param>
         protected async Task ProcessLegacyHeadersAsync(INetworkPeer peer, List<BlockHeader> headers)
         {
+            if (this.isGateway && !peer.IsWhitelisted())
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: Same should be for proven headers payload.

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs [hunk 4]
```diff
@@ -184,6 +200,12 @@ protected override GetHeadersPayload BuildGetHeadersPayload()
         {
             INetworkPeer peer = this.AttachedPeer;
 
+            if (this.isGateway && !peer.IsWhitelisted())
+            {
+                this.logger.LogTrace("Node is a gateway, cannot sync from non whitelisted peer.");
+                return null;
+            }
+
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: later in the code if we are the gateway and peer is whitelisted and it is a C# node we will ask for PH. 
We should ask for normal headers.

Comment (MithrilMan) [reply src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: yes, actually we don't support C# whitelisted node as gateway. If we need to, we can ask for legacy peer but we should implement even a way to communicate using legacy node when the non-gateway c# node has a block to announce to us

Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: why not just do the inverse?

```
if (this.isGateway && peer.IsWhitelisted())
  return base.BuildGetHeadersPayload();
```
And bellow remove any whitelisting reference

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs [hunk 5]
```diff
@@ -214,6 +236,18 @@ protected override GetHeadersPayload BuildGetHeadersPayload()
             }
         }
 
+        /// <inheritdoc />
+        protected override Task ProcessHeadersAsync(INetworkPeer peer, List<BlockHeader> headers)
+        {
+            if (this.isGateway)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: If we are a gateway we just don't processes PH

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs (hunk 3 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: I suggest here the same handle gateway and whitelist and remove whitelist form bellow

File: src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs (hunk 5 above)
Comment (MithrilMan) [reply src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs]: yes this change was to honor that

**Pull request merged**